src/cgame/cg_rocket_draw.cpp

```diff
@@ -2174,6 +2174,111 @@ class PredictedMineEfficiencyElement : public HudElement
 	std::unordered_map<int, std::string> pluralSuffix;
 };
 
+class BarbsHudElement : public HudElement
+{
+public:
+	BarbsHudElement ( const Rocket::Core::String& tag ) :
+	HudElement ( tag, ELEMENT_ALIENS ),
+	numBarbs( 0 ),
+	regenerationInterval ( 0 ),
+	t0 ( 0 ),
+	offset ( 0 ) {}
+
+	void OnAttributeChange( const Rocket::Core::AttributeNameList& changed_attributes )
+	{
+		HudElement::OnAttributeChange( changed_attributes );
+		if ( changed_attributes.find( "src" ) != changed_attributes.end() )
+		{
+			src = GetAttribute<Rocket::Core::String>( "src", "" );
+			Com_sprintf( base, sizeof( base ), "<img class='barbs' src='%s' />", src.CString() );
+		}
+	}
+
+	void DoOnUpdate()
+	{
+		int barbs = cg.snap->ps.ammo;
+		int interval = GetInterval();
+
+		// start regenerating barb now
+		if ( barbs > numBarbs || ( barbs<numBarbs && numBarbs == BG_Weapon( WP_ALEVEL3_UPG )->maxAmmo ) )
+		{
+			t0 = cg.time;
+			// sin(-pi/2) is minimal
+			offset = -M_PI_2;
+		}
+		// change regeneration speed
+		else if ( interval != regenerationInterval )
+		{
+			t0 = cg.time;
+			offset = asin( GetSin() ) - M_PI_2;
+			regenerationInterval = interval;
+		}
+		numBarbs = barbs;
+	}
+
+	void DoOnRender()
+	{
+		char rml[ MAX_STRING_CHARS ] = { 0 };
+
+		// draw existing barbs
+		if ( numBarbs > 0 )
+		{
+			for ( int i=0; i<numBarbs; i++ )
+			{
+				Q_strcat( rml, sizeof( rml ), base );
+			}
+		}
+
+		// draw regenerating barb
+		if ( numBarbs < BG_Weapon( WP_ALEVEL3_UPG )->maxAmmo )
+		{
+			// normalize
+			float opacity = GetSin() / 3 + 0.5;
+			Q_strcat( rml, sizeof( rml ), va( "<img class='barbs' src='%s' style='opacity: %s; color: grey;' />", src.CString(), va( "%f", opacity) ) );
+		}
+
+		SetInnerRML( rml );
```

Comment:
Currently this will set RML each time, which is quite expensive. Setting RML will remove all children from the parent and then re-add them, which means you will need to re-layout everything. You should only set RML once, and then you should store a reference to each of the img elements, and then call element->SetProperty("opacity", va("%f", opacity)) to modify the look.